File: hadoop_ocl_link_test/src/gsod/DataSet.java
package gsod;

public class DataSet {

	public final static int YEAR_START = 15-1;
	public final static int YEAR_END = 18-1;

	public final static int TEMP_START = 25;
	public final static int TEMP_END = 30;

	public final static int MAX_START = 103;
	public final static int MAX_END = 108;

	public final static int MISSING = 9999;

	public static String getYear(String line) {
		return line.substring(YEAR_START, YEAR_END + 1);
	}

	public static int getTemp(String line) {
		return getInt(line, TEMP_START, TEMP_END);
	}

	public static int getMax(String line) {
		return getInt(line, MAX_START, MAX_END);
	}

	private static int getInt(String line, final int START, final int END) {
		line = line.replace('*', ' ');
		return (int)(Double.parseDouble(line.substring(START, END + 1)) * 10);
	}

	public static void main(String[] args) {
		String line = "010010 99999  20081220    24.6 24    18.5 24   997.8 24   996.6 24    5.7  6   15.6 24   27.2  999.9    29.8*   20.7*  0.01G 999.9  001000";
		System.out.println(DataSet.getYear(line));
		System.out.println(DataSet.getTemp(line));
		System.out.println(DataSet.getMax(line));
	}

}

File: hadoop_ocl_link_test/src/gsod_cpp/DataSet.cpp
#include <string>
#include <iostream>
#include <stdlib.h>

namespace gsod {

class DataSet {
public:
	static const int YEAR_START = 14;
	static const int YEAR_DIFF = 4;

	static const int TEMP_START = 25;
	static const int TEMP_DIFF = 6;

	static const int MAX_START = 103;
	static const int MAX_DIFF = 6;

	static const int MISSING = 9999;

	static std::string getYear(std::string line);

	static int getTemp(std::string line);

	static int getMax(std::string line);

private:
	static int getInt(std::string line, const int START, const int DIFF);

};

}

using namespace gsod;

std::string DataSet::getYear(std::string line) {
	return line.substr(YEAR_START, YEAR_DIFF);
}

int DataSet::getTemp(std::string line) {
	return getInt(line, DataSet::TEMP_START, DataSet::TEMP_DIFF);
}

int DataSet::getMax(std::string line) {
	return getInt(line, DataSet::MAX_START, DataSet::MAX_DIFF);
}

int DataSet::getInt(std::string line, const int START, const int DIFF) {
	return (int) (strtod(line.substr(START, DIFF).data(), NULL) * 10);
}

#ifdef TESTING

int main(int argc, char* argv[]) {
	std::string line("654420 99999  20070703    73.0  5    70.6  5  9999.9  0  9999.9  0    6.4  5    3.8  5    8.0  999.9    77.0*   71.6*  0.04A 999.9  010000");
	std::cout << line << std::endl;

	std::cout << "Year: " << DataSet::getYear(line) << std::endl;
	std::cout << "Temp: " << DataSet::getTemp(line) << std::endl;
	std::cout << "Max: " << DataSet::getMax(line) << std::endl;

	return 0;
}

#endif /* TESTING */

